test_binary_generation/generate_raw_ltoir.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+import argparse
+import pathlib
+import subprocess
+import sys
+
+from cuda import nvrtc
+
+# Magic number found at the start of an LTO-IR file
+LTOIR_MAGIC = 0x7F4E43ED
+
+
+def check(args):
+    """
+    Abort and print an error message in the presence of an error result.
+
+    Otherwise:
+    - Return None if there were no more arguments,
+    - Return the singular argument if there was only one further argument,
+    - Return the tuple of arguments if multiple followed.
+    """
+
+    result, *args = args
+    value = result.value
+
+    if value:
+        error_string = check(nvrtc.nvrtcGetErrorString(result)).decode()
+        msg = f"NVRTC error, code {value}: {error_string}"
+        print(msg, file=sys.stderr)
+        sys.exit(1)
+
+    if len(args) == 0:
+        return None
+    elif len(args) == 1:
+        return args[0]
+    else:
+        return args
+
+
+def determine_include_path():
+    # Inspired by the logic in FindCUDA.cmake. We need the CUDA include path
+    # because NVRTC doesn't add it by default, and we can compile a much
+    # broader set of test files if the CUDA headers are available.
+
+    # We get NVCC to tell us the location of the CUDA Toolkit.
+
+    cmd = ["nvcc", "-v", "__dummy"]
+    cp = subprocess.run(cmd, capture_output=True)
+
+    rc = cp.returncode
+    if rc != 1:
+        print(f"Unexpected return code ({rc}) from `nvcc -v`. Expected 1.")
+        return None
+
+    output = cp.stderr.decode()
+    lines = output.splitlines()
+
+    top_lines = [line for line in lines if line.startswith("#$ TOP=")]
```

Comment:
I was a bit nervous about the parsing of `#$ INCLUDES` - I don't know how complicated that can get; whereas parsing `TOP`, I could just use equivalent logic to `FindCUDA.cmake`.

Reply:
I guess with this approach, we are assuming that the only include paths for cuda headers is "$CUDA_HOME/include". In my parsed `$INCLUDES` variable, this is also the only include folder. 

However, there's a reason why they set additional `$INCLUDES` variable so that we are not making such assumption in case the CTK wants to be flexible in including it's headers.

---

But, since `FindCUDA.cmake` is a well tested script in finding CTK, I'm also ok with the current approach and lift the discussion above with @robertmaynard

Reply:
Don't use the approach in `FindCUDA.cmake` as it isn't correct, and still exists since the module is deprecated and not mainatined  anymore.

Specifically `TOP` only relates to the `nvcc` binary that is launched and has no bearing on what includes should be used. The includes are controlled by the target cpu compilation arch and can change based on what platform you are targetting.

So if we use a CUDA setup on an x86 host with support for x86 and aarch64 compilation we can see the changes.

**x86 to x86 compilation:**
```
nvcc -v __dummy
...
#$ _HERE_=/usr/local/cuda/bin
....
#$ _TARGET_DIR_=targets/x86_64-linux
#$ TOP=/usr/local/cuda/bin/..
...
#$ INCLUDES="-I/usr/local/cuda/bin/../targets/x86_64-linux/include"  
#$ LIBRARIES=  "-L/usr/local/cuda/bin/../targets/x86_64-linux/lib/stubs" "-L/usr/local/cuda/bin/../targets/x86_64-linux/lib"
```

**x86 to aarch64 compilation:**
```
nvcc -v __dummy -ccbin aarch64-linux-gnu-g++-10
...
#$ _HERE_=/usr/local/cuda/bin
....
#$ _TARGET_DIR_=targets/sbsa-linux
#$ TOP=/usr/local/cuda/bin/..
...
#$ INCLUDES="-I/usr/local/cuda/bin/../targets/sbsa-linux/include"  
#$ LIBRARIES=  "-L/usr/local/cuda/bin/../targets/sbsa-linux/lib/stubs" "-L/usr/local/cuda/bin/../targets/sbsa-linux/lib"
```

To properly handle things like conda cross compilation enviornments ( where -ccbin= is passed implicitly ) you need to scrape `INCLUDES` for proper handling of differernt `target` values. Likewise some CUDA Toolkit installs place headers into multiple locations, and therefore you should expect `INCLUDES` to contain multiple paths.

What CMake currently does is iterate over `INCLUDES` and strips any leading `-I` or `-isystem` to get the set of include paths to search.

Reply:
For future reference, [FindCUDAToolkit](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html) is the modern replacement for FindCUDA.

Reply:
Thanks @vyasr and @robertmaynard for the pointers - I've now updated the logic to follow what `FindCUDAToolkit` does, and I can see how it's more robust (cc @isVoid ).

Reply:
Looks great